Merge Android port with Mac and Win ports

kollivier opened the pull request:
One codebase to run them all!

kollivier commented on src/main.py:
```diff
@@ -395,7 +352,7 @@ def create_kolibri_window(self, url):
         history_menu = pew.ui.PEWMenu(_('History'))
         history_menu.add(_('Back'), handler=window.on_back, shortcut=PEWShortcut('[', modifiers=['CTRL']))
-        history_menu.add(_('Forward'), handler=window.on_forward, shortcut=PEWShortcut(']', modifiers=['CTRL']))
+        history_menu.add(_('Forward'), handler=window.on_forward, shortcut=PEWShortcut(']fclas', modifiers=['CTRL']))
```
Ack, thanks! Sometimes PyCharm does weird things with focus, and this sort of problem happens. I usually catch these, but I guess due to looking at the large-ish PR my eyes glossed over this. 😓

kollivier commented on src/main.py:
```diff
@@ -352,13 +301,21 @@ def setUp(self):
+
+        if pew.ui.platform == "android":
+            logging.info("Starting kolibri as Android service...")
 
-        logging.info("Preparing to start Kolibri server...")
-        self.server_thread = pew.ui.PEWThread(target=start_django, kwargs={'port': self.port})
-        self.server_thread.daemon = True
-        self.server_thread.start()
+            from platforms.android.utils import start_service
+            start_service("kolibri", dict(os.environ))
+        else:
+            if self.server_thread:
+                del self.server_thread
```
Yes, good point, I was treating this as a next step, but it makes sense to go ahead and do this now in order to simplify the startup code as well.

kollivier replied:
@dylanmccall On second thought, going to handle this as a subsequent PR after merging this one, so we can finish up the review on this, since it's become quite meaty. :)

kollivier replied:
Issue filed here: https://github.com/learningequality/kolibri-installer-mac/issues/73

kollivier commented on src/config.py:
```diff
@@ -0,0 +1,2 @@
+KOLIBRI_PORT = 8080
```
I'm not personally attached to it, and actually would prefer that we move over to the approach you're working on at https://github.com/learningequality/kolibri/pull/7714. This is the port Android was using so I just kept it during the migration.

kollivier replied:
Can switch it back to 5000 or some other port in the meantime as well.

kollivier replied:
I think one option would be to trying one or more default ports first, that Kolibri is aware of, and then choosing another random port only if those fail. This would allow things to work except in the case of "all default ports used + zeroconf detection not working" case, which should hopefully be fairly uncommon.

jamalex replied:
Yeah, that could be a good compromise!

kollivier replied:
I've switched the port to 5000, and will add the ability to check for open ports in a subsequent PR.

kollivier replied:
Filed an issue here: https://github.com/learningequality/kolibri-installer-mac/issues/74

jamalex commented on src/kolibri_tools/utils.py:
```diff
@@ -0,0 +1,76 @@
+    return get_content_storage_file_path(filename)
+
+
+def automatic_provisiondevice():
```
Is this something we'd perhaps want in Kolibri proper, @rtibbles? Then it could be used for non-app installations as well.

rtibbles replied:
Yes, I think so - also means that this functionality should be less prone to breakage if something is changed in provision device.

kollivier replied:
Going to file an issue for this in the Kolibri repo, and leave it as a next step.

kollivier replied:
Issue filed! https://github.com/learningequality/kolibri/issues/7803

jamalex commented on src/platforms/android/service.py:
```diff
@@ -0,0 +1,83 @@
+
+    share_by_intent(**kwargs)
+
+    return "<html><body style='background: white;'>OK, boomer</body></html>"
```
Hah! Now the boomers will `git blame` you for this!

kollivier replied:
Or they may now think it's a deepcut anime reference! ;-)

https://en.wikipedia.org/wiki/List_of_Bubblegum_Crisis_characters#Boomers

kollivier replied:
As a just-in-case provision, I changed the wording to "Success" instead. :)

jamalex reviewed (commented):
Added a few notes. Definitely seems valuable to reconcile/integrate like this. My main concern is that with the shared cross-platform code, but only being able to run/build the code for one's own platform, there's room for changing something to get it work for your own platform that breaks it for other platforms, and we don't have automated test coverage here. Not sure how best to mitigate that.

kollivier commented on Makefile:
```diff
@@ -0,0 +1,38 @@
+	./scripts/rundocker.sh
+
+softbuild: project_info.json
+	pew build $(pew_release_flag) android ${ARCH}
```
Agreed. I initially tried this, but had some issues with path resolution, and so I figured I'd get it working first and then consult with @DXCanas on how to move them out of the top level.

kollivier replied:
Also, regarding your other comment:

> My main concern is that with the shared cross-platform code, but only being able to run/build the code for one's own platform, there's room for changing something to get it work for your own platform that breaks it for other platforms, and we don't have automated test coverage here. Not sure how best to mitigate that.

For GUI applications, I've found there's no easy way around doing manual testing of releases on each platform, as automated testing is often very platform-specific and costly to implement, and is often fragile. We could add some sort of issue template with a set of "This PR has been manually tested on [Win | Mac | Android]" checkboxes as a way to ensure we're doing testing on all supported platforms. If someone doesn't have all the environments set up, then maybe create a test scenario (a good practice generally) and put in a request for QA to do some testing of the builds.

kollivier commented on src/kolibri_tools/initialization.py:
```diff
@@ -0,0 +1,26 @@
+    os.environ["DJANGO_SETTINGS_MODULE"] = "kolibri_tools.django_app_settings"
+
+    # TODO: before shipping the app, make this contingent on debug vs production mode
+    os.environ["KOLIBRI_RUN_MODE"] = "pew-dev"
```
Do we want to remove this entirely, or set it to something else? I think Android is the only platform where the user cannot set `KOLIBRI_RUN_MODE` themselves easily. 

Another possibility would be to use a versioning scheme where official releases can be clearly distinguished. (e.g. using alpha or beta tags in non-production builds), and set KOLIBRI_RUN_MODE in the app when those tags are found in the version string. This would allow us to strip out pre-release builds from stats regardless of the machine they were run on.

jamalex replied:
We should remove it entirely, yeah. We do already detect and automatically mark a "dev" runmode on the pingbacks for non-production releases, so we can just lean on that.

kollivier commented on build_tools/version.py:
```diff
@@ -0,0 +1,90 @@
+    Returns the build number for the apk. This is the mechanism used to understand whether one
+    build is newer than another. Uses buildkite build number with time as local dev backup
+    """
+    return os.getenv('BUILDKITE_BUILD_NUMBER', datetime.now().strftime('%y%m%d%H%M'))
```
Sorry, does look like I missed that, I just pushed an update that should resolve it! (Untested as this repo isn't yet setup for Android BuildKite builds yet.)

kollivier commented on README.md:
```diff
@@ -1,42 +1,82 @@
-Run pipenv to download and install dependencies:
+## Getting Started
+
+**IMPORTANT:** On Windows, there is a bug with PyInstaller and the latest virutalenv.
```
Thanks, pushed a fix!

kollivier commented:
@dylanmccall @jamalex I think I've addressed all the concerns raised, and / or filed an issue for followup. I'm not going to hit the merge button on a Friday afternoon, but hoping to do so on Monday as we're hoping to do an app session on Tuesday and want to have the initial implementation merged in by then. Please let me know if you have any objections!
